src/runtime/runtime_rp2040.go

```diff
@@ -43,6 +50,284 @@ func sleepTicks(d timeUnit) {
+	for multicore_fifo_rvalid() {
+		rp.SIO.FIFO_RD.Get()
+	}
+}
+
+func multicore_fifo_push_blocking(data uint32) {
+	for !multicore_fifo_wready() {
+	}
+	rp.SIO.FIFO_WR.Set(data)
+	arm.Asm("sev")
+}
+
+func multicore_fifo_pop_blocking() uint32 {
+	for !multicore_fifo_rvalid() {
+		arm.Asm("wfe")
+	}
+
+	return rp.SIO.FIFO_RD.Get()
+}
+
+// Value used to communicate between the GC core and the other (paused) cores.
+var gcSignalWait volatile.Register8
+
+// The GC interrupted this core for the stop-the-world phase.
+// This function handles that, and only returns after the stop-the-world phase
+// ended.
+func gcInterruptHandler(hartID uint32) {
+	// Let the GC know we're ready.
+	gcScanState.Add(1)
+	arm.Asm("sev")
+
+	// Wait until we get a signal to start scanning.
+	for gcSignalWait.Get() == 0 {
+		arm.Asm("wfe")
+	}
+	gcSignalWait.Set(0)
+
+	// Scan the stack(s) of this core.
+	scanCurrentStack()
+	if !task.OnSystemStack() {
+		// Mark system stack.
+		markRoots(task.SystemStack(), coreStackTop(hartID))
+	}
+
+	// Signal we've finished scanning.
+	gcScanState.Store(1)
+	arm.Asm("sev")
+
+	// Wait until we get a signal that the stop-the-world phase has ended.
+	for gcSignalWait.Get() == 0 {
+		arm.Asm("wfe")
+	}
+	gcSignalWait.Set(0)
+
+	// Signal we received the signal and are going to exit the interrupt.
+	gcScanState.Add(1)
+	arm.Asm("sev")
+}
+
+// Pause the given core by sending it an interrupt.
+func gcPauseCore(core uint32) {
+	rp.SIO.FIFO_WR.Set(1)
+}
+
+// Signal the given core that it can resume one step.
+// This is called twice after gcPauseCore: the first time to scan the stack of
+// the core, and the second time to end the stop-the-world phase.
+func gcSignalCore(core uint32) {
+	gcSignalWait.Set(1)
+	arm.Asm("sev")
+}
+
+// Returns the stack top (highest address) of the system stack of the given
+// core.
+func coreStackTop(core uint32) uintptr {
+	switch core {
+	case 0:
+		return uintptr(unsafe.Pointer(&stackTopSymbol))
+	case 1:
+		return uintptr(unsafe.Pointer(&stack1TopSymbol))
+	default:
+		runtimePanic("unexpected core")
+		return 0
+	}
+}
+
+// These spinlocks are needed by the runtime.
+var (
+	printLock     = spinLock{id: 0}
+	schedulerLock = spinLock{id: 1}
+	atomicsLock   = spinLock{id: 2}
+	futexLock     = spinLock{id: 3}
+)
+
+// A hardware spinlock, one of the 32 spinlocks defined in the SIO peripheral.
+type spinLock struct {
+	id uint8
+}
+
+// Return the spinlock register: rp.SIO.SPINLOCKx
+func (l *spinLock) spinlock() *volatile.Register32 {
+	return (*volatile.Register32)(unsafe.Add(unsafe.Pointer(&rp.SIO.SPINLOCK0), l.id*4))
+}
+
+func (l *spinLock) Lock() {
+	// Wait for the lock to be available.
+	spinlock := l.spinlock()
+	for spinlock.Get() == 0 {
+		// TODO: use wfe and send an event when unlocking so the CPU can go to
+		// sleep while waiting for the lock.
+		// Unfortunately when doing that, time.Sleep() seems to hang somewhere.
+		// This needs some debugging to figure out.
```

Comment:
This could hide some other subtle error that affects not just `time.Sleep`. It seems risky to enable the multicore scheduler by default in v0.38.0 with so little bake time.

Reply:
"the voice of reason"? :smiley_cat:

This should build with multicore I think:
 
```
tinygo flash -target=pico -scheduler=cores -ldflags="--defsym=__num_stacks=2" path/to/code
```

Perhaps we should consider it experimental for this release?

Reply:
Another implication of changing the default to multicore, is that default power use will be higher with the second core active. :zap:

Reply:
I don't think power consumption will be affected much: when one of the cores has nothing to do it will just go to sleep. But I didn't measure this. (Should be easy to do by compiling with -scheduler=tasks and -scheduler=cores and measuring the difference).

Reply:
So, yeah, I did not investigate that problem with wfe/sev in the spinlock. I would need to attach a debugger to see where it gets stuck.

We could disable multicore support for now, and leave it for people who want to test it. It's possible there's a bug I didn't discover yet. That comment is just there since I had to back out wfe/sev at the last minute and wanted to make sure the next person knows there's an opportunity for optimization there.

Reply:
> That comment is just there since I had to back out wfe/sev at the last minute and wanted to make sure the next person knows there's an opportunity for optimization there.

That was what I was trying to say. There is not a known bug here, this is a chance to optimize further.

Reply:
> I don't think power consumption will be affected much: when one of the cores has nothing to do it will just go to sleep.

So it might use some extra power initially, but then would go down to minimal use on that core?

Reply:
So the way it works (at least how I understand it), is that both cores are always running. The second core is just asleep on reset. With multicore support, it is pulled out of this sleep state and starts the scheduler - which in many cases will put it right back to sleep. So yeah it spends a little time configuring stuff and then goes back to sleep.
...but again, I didn't measure anything so I'm only speculating how it *should* work.

Reply:
> So yeah it spends a little time configuring stuff and then goes back to sleep.

That is what all of the docs and references I could find are claiming about power use. So it would appear to be fine as it is from that point of view.

Reply:
> We could disable multicore support for now, and leave it for people who want to test it. It's possible there's a bug I didn't discover yet.

As much as this feature excites me, I think @eliasnaur is probably correct that we should not make it the default yet.

Reply:
@aykevl has a better idea of the robustness of the multi-core support than I; I merely pointed out the timing.

I'm also very excited about this code, and I would be comfortable making multi-core default, if the release notes spelled out instructions for switching back to the single-core scheduler. It's just a `--scheduler=...` argument, right? Assuming `__num_stacks=2` is benign.

Reply:
I had already confirmed that `-scheduler=tasks` does what you would expect and runs on single core.

I assume that the extra stack space is wasted in that case, so I also tested this command (note I missed the `extldflags` in my original comment):

```
tinygo flash -target nano-rp2040 -ldflags="--extldflags '--defsym=__num_stacks=1'" -stack-size 8kb -scheduler tasks /path/to/my/code
```

This also worked correctly, and according to `-size full` did save some bytes.

Based on this, I think we can merge this PR as it is, and just add the docs to fallback to single core.